File: doc-examples/example-kotlin/src/test/kotlin/example/BookMethodsTest.kt
package example

import io.micronaut.el.CompiledELContext
import jakarta.el.ELManager
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class BookMethodsTest {

    private val factory = ELManager.getExpressionFactory()
    private val context = CompiledELContext().setBean("book", Book("Jakarta EL", "reference", 20.0))

    private fun evaluate(expression: String): Any? =
        factory.createValueExpression(context, expression, Any::class.java).getValue(context)

    @Test
    fun theContributedMethodsAreCallableFromAnExpressionParsedAtRuntime() {
        assertEquals("Jakarta EL", evaluate("\${book.title()}"))
        assertEquals(18.0, evaluate("\${book.discounted(10)}"))
        assertEquals("\$18.0", evaluate("\${book.label('\$', 10)}"))
        assertEquals("Jakarta EL a-b", evaluate("\${book.tagged('-', 'a', 'b')}"))
        assertEquals(7L, evaluate("\${Math.abs(-7)}"))
        assertEquals("HI!", evaluate("\${fmt:shout('hi')}"))
    }

    @Test
    fun aLambdaReachesTheApplicationInterfaceWithoutAProxy() {
        assertEquals("Jakarta EL", evaluate("\${book.summarised(b -> b.title())}"))
    }
}

File: doc-examples/example-python/src/test/python/example/BookMethodsTest.py
from java.lang import Object
from java.util import List
from micronaut.el import CompiledELContext, CompiledExpressionFactory, ContributedELMethodExecutor
from micronaut.el.interpreter import InterpretingELExpressionParser
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Disabled, Test

from example.Book import Book
from example.BookMethods import BookMethods


@MicronautTest
class BookMethodsTest:

    def __init__(self):
        # TODO(python): the contributor is also named in META-INF/services/io.micronaut.el.ELMethodContributor, but
        # ELContributions loads the services once per JVM and keeps the registrations: the lambdas BookMethods
        # registered belong to the GraalPy context of the first test class that touched the factory and are unusable
        # here ("Context execution was cancelled"), so this test hands the parser the contributor explicitly, the way
        # the guide shows for leaving executors out.
        self.factory = CompiledExpressionFactory(List.of(),
            InterpretingELExpressionParser(List.of(ContributedELMethodExecutor(List.of(BookMethods())))))
        self.context = CompiledELContext().setBean("book", Book("Jakarta EL", "reference", 20.0))

    def evaluate(self, expression: str):
        return self.factory.createValueExpression(self.context, expression, Object).getValue(self.context)

    @Test
    def test_the_contributed_methods_are_callable_from_an_expression_parsed_at_runtime(self):
        assert self.evaluate("${book.title()}") == "Jakarta EL"
        assert self.evaluate("${book.discounted(10)}") == 18.0
        assert self.evaluate("${book.label('$', 10)}") == "$18.0"
        assert self.evaluate("${book.tagged('-', 'a', 'b')}") == "Jakarta EL a-b"
        assert self.evaluate("${Math.abs(-7)}") == 7
        assert self.evaluate("${fmt:shout('hi')}") == "HI!"

    # TODO(python): the coercion of a lambda expression to the application's functional interface is resolved through
    # the JVM-wide ELContributions registry (ELSupport.functionalInterface), not through the executor this test hands
    # the parser: that registry holds the BookMethods instance the service loader created in the GraalPy context of
    # whichever test class touched it first, so the outcome depends on the test class order ("Context execution was
    # cancelled" on CI). Runs green in an application, which has one context.
    @Disabled("TODO(python): the lambda coercion goes through the JVM-wide contributor registry, whose Python contributor belongs to another test class's context")
    @Test
    def test_a_lambda_reaches_the_application_interface_without_a_proxy(self):
        assert self.evaluate("${book.summarised(b -> b.title())}") == "Jakarta EL"
